### PR TITLE
Fix deferrable mode for DataflowTemplatedJobStartOperator and DataflowStartFlexTemplateOperator

### PR DESCRIPTION
**Pull request by e-galan**
- Adds new methods to `DataflowHook` to be used for deferrable mode. The methods will start a dataflow job, and then exit returning the job data.
- Add statements to push job_id to XCOM in both sync and deferrable runs.
- Update unit tests
- Small refactoring of data types and function parameters

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/google/cloud/operators/dataflow.py [hunk 1]
```diff
@@ -680,29 +681,34 @@ def set_current_job(current_job):
...
-                "While reading job object after template execution error occurred. Job object has no id."
-            )
-
-        if not self.deferrable:
-            return job_id
-
-        context["ti"].xcom_push(key="job_id", value=job_id)
-
+        job_id = self.hook.extract_job_id(self.job)
+        DataflowJobLink.persist(self, context, self.project_id, self.location, job_id)
```
Comment (moiseenkov) [inline airflow/providers/google/cloud/operators/dataflow.py:711]: Could you please also add this to the `execute_complete()` method?

File: airflow/providers/google/cloud/operators/dataflow.py [hunk 2]
```diff
@@ -870,32 +882,35 @@ def set_current_job(current_job):
...
-        if job_id is None:
-            raise AirflowException(
-                "While reading job object after template execution error occurred. Job object has no id."
-            )
-
-        if not self.deferrable:
-            return self.job
-
+        job_id = self.hook.extract_job_id(self.job)
+        DataflowJobLink.persist(self, context, self.project_id, self.location, job_id)
```
Comment (moiseenkov) [inline airflow/providers/google/cloud/operators/dataflow.py:902]: The same here

File: airflow/providers/google/cloud/operators/dataflow.py (hunk 1 above)
Comment (MaksYermak) [inline airflow/providers/google/cloud/operators/dataflow.py:711]: What about link for non deferrable mode?

**Review by moiseenkov — APPROVED** [review]

File: airflow/providers/google/cloud/operators/dataflow.py (hunk 2 above)
Comment (MaksYermak) [inline airflow/providers/google/cloud/operators/dataflow.py:902]: The same maybe it makes sense to add a link for non-deff mode.

File: airflow/providers/google/cloud/operators/dataflow.py (hunk 1 above)
Comment (e-galan) [reply airflow/providers/google/cloud/operators/dataflow.py:711]: It's already there, please take a look at `set_current_job()`  function on line 674.

Comment (MaksYermak) [reply airflow/providers/google/cloud/operators/dataflow.py:711]: Got it

**Review by MaksYermak — APPROVED** [review]

**Pull request closed without merging**